Handle broken project / unknown projects in the dependency calculation

There is NPE in the dependency calculation if processed projects are not N4JS projects (just other Eclipse projects) or if  index is broken (e.g. after crash) and there project information is not available.

```
java.lang.reflect.InvocationTargetException
        at org.eclipse.jface.operation.ModalContext.run(ModalContext.java:398)
        at org.eclipse.jface.dialogs.ProgressMonitorDialog.run(ProgressMonitorDialog.java:481)
        at org.eclipse.n4js.ui.handler.LibrariesFixHandler.execute(LibrariesFixHandler.java:95)
        at org.eclipse.ui.internal.handlers.HandlerProxy.execute(HandlerProxy.java:295)
        at org.eclipse.ui.internal.handlers.E4HandlerProxy.execute(E4HandlerProxy.java:90)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:55)
        at org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(InjectorImpl.java:282)
        at org.eclipse.e4.core.internal.di.InjectorImpl.invoke(InjectorImpl.java:264)
        at org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(ContextInjectionFactory.java:132)
        at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.execute(HandlerServiceHandler.java:152)
        at org.eclipse.core.commands.Command.executeWithChecks(Command.java:494)
        at org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(ParameterizedCommand.java:488)
        at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:210)
        at org.eclipse.e4.ui.workbench.renderers.swt.HandledContributionItem.executeItem(HandledContributionItem.java:433)
        at org.eclipse.e4.ui.workbench.renderers.swt.AbstractContributionItem.handleWidgetSelection(AbstractContributionItem.java:454)
        at org.eclipse.e4.ui.workbench.renderers.swt.AbstractContributionItem$3.handleEvent(AbstractContributionItem.java:482)
        at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:84)
        at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4256)
        at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1501)
        at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1524)
        at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1509)
        at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1313)
        at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4080)
        at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3706)
        at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$4.run(PartRenderingEngine.java:1121)
        at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
        at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1022)
        at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:150)
        at org.eclipse.ui.internal.Workbench$5.run(Workbench.java:693)
        at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
        at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:610)
        at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:148)
        at org.eclipse.n4js.product.N4JSApplication.start(N4JSApplication.java:141)
        at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
        at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
        at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
        at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:388)
        at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:243)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:673)
        at org.eclipse.equinox.launcher.Main.basicRun(Main.java:610)
        at org.eclipse.equinox.launcher.Main.run(Main.java:1519)
Caused by: java.lang.NullPointerException
        at org.eclipse.n4js.projectModel.dependencies.DependenciesCollectingUtil.lambda$0(DependenciesCollectingUtil.java:47)
        at java.lang.Iterable.forEach(Iterable.java:75)
        at org.eclipse.n4js.projectModel.dependencies.DependenciesCollectingUtil.updateMissingDependneciesMap(DependenciesCollectingUtil.java:46)
        at org.eclipse.n4js.ui.handler.ProjectDependenciesHelper.calculateDependenciesToInstall(ProjectDependenciesHelper.java:63)
        at org.eclipse.n4js.ui.handler.LibrariesFixHandler.setupWorkspaceDependnecies(LibrariesFixHandler.java:180)
        at org.eclipse.n4js.ui.handler.LibrariesFixHandler$1.run(LibrariesFixHandler.java:88)
        at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:119)
Root exception:
java.lang.NullPointerException
        at org.eclipse.n4js.projectModel.dependencies.DependenciesCollectingUtil.lambda$0(DependenciesCollectingUtil.java:47)
        at java.lang.Iterable.forEach(Iterable.java:75)
        at org.eclipse.n4js.projectModel.dependencies.DependenciesCollectingUtil.updateMissingDependneciesMap(DependenciesCollectingUtil.java:46)
        at org.eclipse.n4js.ui.handler.ProjectDependenciesHelper.calculateDependenciesToInstall(ProjectDependenciesHelper.java:63)
        at org.eclipse.n4js.ui.handler.LibrariesFixHandler.setupWorkspaceDependnecies(LibrariesFixHandler.java:180)
        at org.eclipse.n4js.ui.handler.LibrariesFixHandler$1.run(LibrariesFixHandler.java:88)
        at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:119)
```

--- a/plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/handler/LibrariesFixHandler.java
+++ b/plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/handler/LibrariesFixHandler.java
@@ -1,6 +1,8 @@
 package org.eclipse.n4js.ui.handler;
 
 import static org.eclipse.core.resources.ResourcesPlugin.getWorkspace;
+import static org.eclipse.jface.dialogs.MessageDialog.openError;
+import static org.eclipse.n4js.ui.utils.UIUtils.getDisplay;
 import static org.eclipse.ui.PlatformUI.isWorkbenchRunning;
 
 import java.io.File;
@@ -95,6 +97,12 @@ public class LibrariesFixHandler extends AbstractHandler {
 			dependenciesDialog.run(true, true, iRunnableWithProgress);
 		} catch (InvocationTargetException | InterruptedException | SWTException err) {
 			LOGGER.error("unhandled error while setting up dependencies", err);
+			getDisplay().asyncExec(() -> openError(
+					UIUtils.getShell(),
+					"Setting up external libraries failed.",
+					"Error while setting up external libraries.\n"
+							+ "Please check your Error Log view for the detailed log about the failure.\n" +
+							" (note that autobuild is " + wasAutoBuilding + ")"));
 		}
 
 		return null;

--- a/plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/handler/LibrariesFixHandler.java
+++ b/plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/handler/LibrariesFixHandler.java
@@ -102,7 +102,7 @@ public class LibrariesFixHandler extends AbstractHandler {
 					"Setting up external libraries failed.",
 					"Error while setting up external libraries.\n"
 							+ "Please check your Error Log view for the detailed log about the failure.\n" +
-							" (note that autobuild is " + wasAutoBuilding + ")"));
+							" (note that autobuild is " + getAutobuildSetting() + ")"));
 		}
 
 		return null;

--- a/plugins/org.eclipse.n4js/src/org/eclipse/n4js/projectModel/dependencies/DependenciesCollectingUtil.java
+++ b/plugins/org.eclipse.n4js/src/org/eclipse/n4js/projectModel/dependencies/DependenciesCollectingUtil.java
@@ -44,8 +44,12 @@ public class DependenciesCollectingUtil {
 			Iterable<ProjectDescription> projectDescriptions) {
 		final Set<String> availableProjectsIds = new HashSet<>();
 		projectDescriptions.forEach(pd -> {
-			availableProjectsIds.add(pd.getProjectId());
-			updateFromProjectDescription(versionedPackages, pd);
+			// in case we get non N4JS projects, user docs projects that are not N4JS projects, or something created by
+			// the plugins e.g RemoteSystemsTempFiles (see https://stackoverflow.com/q/3627463/52564 )
+			if (pd != null) {
+				availableProjectsIds.add(pd.getProjectId());
+				updateFromProjectDescription(versionedPackages, pd);
+			}
 		});
 		availableProjectsIds.forEach(versionedPackages::remove);
 	}

